engine/engine_test.go

```diff
@@ -330,6 +342,551 @@ func TestSetDefaultWebsocketDataHandler(t *testing.T) {
 	}
 }
 
+func TestAllExchangeWrappers(t *testing.T) {
+	t.Parallel()
+	cfg := config.GetConfig()
+	err := cfg.LoadConfig("../testdata/configtest.json", true)
+	if err != nil {
+		t.Fatal("load config error", err)
+	}
+	isCITest := os.Getenv("CI_TEST")
+	for i := range cfg.Exchanges {
+		name := strings.ToLower(cfg.Exchanges[i].Name)
+		t.Run(name+" wrapper tests", func(t *testing.T) {
+			t.Parallel()
+			if common.StringDataContains(unsupportedExchangeNames, name) {
+				t.Skipf("skipping unsupported exchange %v", name)
+			}
+			ctx := context.Background()
+			if isCITest == "true" && common.StringDataContains(blockedCIExchanges, name) {
+				// rather than skipping tests where execution is blocked, provide an expired
+				// context, so no executions can take place
+				var cancelFn context.CancelFunc
+				ctx, cancelFn = context.WithTimeout(context.Background(), 0)
+				cancelFn()
+			}
+			exch, assetPairs := setupExchange(ctx, t, name, cfg)
+			executeExchangeWrapperTests(ctx, t, exch, assetPairs)
+		})
+	}
+}
+
+func setupExchange(ctx context.Context, t *testing.T, name string, cfg *config.Config) (exchange.IBotExchange, []assetPair) {
+	t.Helper()
+	em := NewExchangeManager()
+	exch, err := em.NewExchangeByName(name)
+	if err != nil {
+		t.Fatalf("%v %v", name, err)
+	}
+	var exchCfg *config.Exchange
+	exchCfg, err = cfg.GetExchangeConfig(name)
+	if err != nil {
+		t.Fatalf("%v %v", name, err)
+	}
+	exch.SetDefaults()
+	exchCfg.API.AuthenticatedSupport = true
+	exchCfg.API.Credentials.Key = "realKey"
+	exchCfg.API.Credentials.Secret = "realSecret"
+	exchCfg.API.Credentials.ClientID = "realClientID"
+	err = exch.Setup(exchCfg)
+	if err != nil {
+		t.Fatalf("%v %v", name, err)
+	}
+
+	err = exch.UpdateTradablePairs(ctx, true)
+	if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+		t.Fatalf("%v %v", name, err)
+	}
+	b := exch.GetBase()
+	assets := b.CurrencyPairs.GetAssetTypes(false)
+	if len(assets) == 0 {
+		t.Fatalf("exchange '%v' has not assets", name)
+	}
+	for j := range assets {
+		err = b.CurrencyPairs.SetAssetEnabled(assets[j], true)
+		if err != nil && !errors.Is(err, currency.ErrAssetAlreadyEnabled) {
+			t.Fatalf("%v %v", name, err)
+		}
+	}
+
+	// Add +1 to len to verify that exchanges can handle requests with unset pairs and assets
+	assetPairs := make([]assetPair, 0, len(assets)+1)
+	for j := range assets {
+		var pairs currency.Pairs
+		pairs, err = b.CurrencyPairs.GetPairs(assets[j], true)
+		if err != nil {
+			t.Fatalf("%v %v", name, err)
+		}
+		var p currency.Pair
+		if len(pairs) == 0 {
+			pairs, err = b.CurrencyPairs.GetPairs(assets[j], false)
+			if err != nil {
+				t.Fatalf("%v GetPairs %v %v", name, err, assets[j])
+			}
+			p, err = getPairFromPairs(t, pairs)
+			if err != nil {
+				// if there aren't pairs to grab, don't test asset
```

Comment:
You'll need to provide more specifics on what you want here in order for me to action it. I don't see it as making sense to only specifically seek certain exchanges here when this can impact all exchanges. In the meantime, what I'll do instead of look for a specific error instead of a catch-all continue